pkg/apis/messaging/v1alpha1/channel_defaults.go

```diff
@@ -0,0 +1,36 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"context"
+	eventingduckv1alpha1 "github.com/knative/eventing/pkg/apis/duck/v1alpha1"
+)
+
+func (c *Channel) SetDefaults(ctx context.Context) {
+	if c != nil && c.Spec.ChannelTemplate == nil {
+		// The singleton may not have been set, if so ignore it and validation will reject the
+		// Channel.
+		if cd := eventingduckv1alpha1.ChannelDefaulterSingleton; cd != nil {
```

Comment:
Shouldn't this be in `ChannelSpec.SetDefaults`?

Reply:
I thought of that but we need the namespace of the Channel. The previous provisioner default mechanism needed the Channel instance, so it is also done within the Channel.SetDefaults. 
Do you know if I can change the contract of ChannelSpec.SetDefaults and also pass a namespace? or will I mess up with some autogen callers?